### PR TITLE
Replace linked lists with array lists

### PR DESCRIPTION
**Pull request by pkoenig10**
## Before this PR
AtlasDB uses linked lists in a couple places.

I noticed this while analyzing a heap dump from an OOM of one of our highest volume services due to a request that read a large amount of data. This service was running with 4GB of heap memory and `LinkedListMultimap$Node` has one of the highest amounts of shallow heap at 272 MB.

Screen Shot 2022-08-12 at 11 16 52 AM

## After this PR
AtlasDB uses array lists instead of linked lists.

There is almost no good reason to use linked lists. Linked lists are much worse for CPU caches and require more memory allocation because of the need to allocate nodes. The amortized constant append costs of an array list are sufficient in almost every case.

### CODE REVIEW

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java [hunk 1]
```diff
@@ -669,17 +670,10 @@ private Map<Cell, Value> getAllCellsForRows(
...
-
-            result.putAll(partialResult);
-
-            query = KeyedStream.stream(Multimaps.asMap(partialResult))
-                    .map((row, cells) -> keyPredicate(row, getNextLexicographicalSlicePredicate(cells)))
-                    .values()
+            query = EntryStream.of(getForKeyPredicates(host, tableRef, query, startTs))
+                    .filterValues(cells -> !cells.isEmpty())
+                    .peekKeyValue(result::putAll)
+                    .mapKeyValue((row, cells) -> keyPredicate(row, getNextLexicographicalSlicePredicate(cells)))
```
Comment (pkoenig10) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java:676]: Small drive-by clean up to avoid a copy

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java [hunk 2]
```diff
@@ -660,7 +661,7 @@ private Map<Cell, Value> getAllCellsForRows(
             final CassandraServer host, final TableReference tableRef, final List<byte[]> rows, final long startTs)
             throws Exception {
 
-        final ListMultimap<ByteBuffer, ColumnOrSuperColumn> result = LinkedListMultimap.create();
+        ListMultimap<ByteBuffer, ColumnOrSuperColumn> result = ArrayListMultimap.create();
```
Comment (schlosna) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java:664]: Do we have any metrics on key to value cardinality here? I'm wondering if we'll bloat with the default ArrayList capacity of 10 entries once we add a single element, in which case we might want to provide a value collection factory.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java [hunk 3]
```diff
@@ -1349,9 +1348,7 @@ private <T> ListenableFuture<Collection<Map.Entry<Cell, T>>> getWithPostFilterin
 
         getCounter(AtlasDbMetricNames.SNAPSHOT_TRANSACTION_CELLS_READ, tableRef).inc(rawResults.size());
 
-        // LinkedList is chosen for fast append operation since we just add to this collection.
```
Comment (schlosna) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java:1352]: Lies 

https://stuartmarks.wordpress.com/2015/12/18/some-java-list-benchmarks/

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java [hunk 4]
```diff
@@ -1349,9 +1348,7 @@ private <T> ListenableFuture<Collection<Map.Entry<Cell, T>>> getWithPostFilterin
...
-        @SuppressWarnings("JdkObsolete")
-        Collection<Map.Entry<Cell, T>> resultsAccumulator = new LinkedList<>();
```
Comment (schlosna) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java:1354]: I'm guessing we haven't enabled JdkObsolete error prone check due to the heavy `Sorted*` interface usages, but that would have flagged this https://errorprone.info/bugpattern/JdkObsolete

**Review by schlosna — COMMENTED** [review]
Overall looks good, one question on multimap values collection to consider

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java (hunk 4 above)
Comment (pkoenig10) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java:1354]: I think we did do this in https://github.com/palantir/atlasdb/pull/5682 (otherwise there probably wouldn't be the suppression here).

The `JdkObsolete` check has lots of exceptions for various things. I assume the remaining uses just fall under those exceptions.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java (hunk 2 above)
Comment (pkoenig10) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java:664]: Good call, I think we can probably just use the `ArrayListMultimap.create` method that takes an `expectedKeys` and `expectedValues` and use 1 for `expectedValues`:
- We're very likely to have one value (client don't often query for keys that don't exist)
- It avoids copying in the single value case, which is fairly common
- We add the values for a given key all at once, so this should require additional growth because the initial capacity is small

**Review by schlosna — APPROVED** [review]

**Review by mdaudali — APPROVED** [review]
lgtm! Thanks for the contribution @pkoenig10 :)
